File: Commons/Math/src/main/java/com/greentree/commons/math/MathLine3D.java
package com.greentree.commons.math;

import com.greentree.commons.math.vector.AbstractVector3f;
import com.greentree.commons.math.vector.Vector3f;

import java.util.Objects;

public class MathLine3D {

    private AbstractVector3f vector;
    private final AbstractVector3f point;

    public MathLine3D(AbstractVector3f point, AbstractVector3f vector) {
        this.point = new Vector3f(point);
        this.vector = new Vector3f(vector);
        if (this.vector.x() < 0)
            this.vector = this.vector.times(-1);
        this.vector.normalize(1f);
    }

    public AbstractVector3f p1() {
        return vector;
    }

    public AbstractVector3f p2() {
        return vector.plus(point);
    }

    @Override
    public int hashCode() {
        return Objects.hash(point, vector);
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        MathLine3D other = (MathLine3D) obj;
        return Objects.equals(point, other.point) && Objects.equals(vector, other.vector);
    }

    @Override
    public String toString() {
        return "MathLine3D [vector=" + vector + ", point=" + point + "]";
    }

}

File: Commons/Math/src/main/java/com/greentree/commons/math/MathLine3D.kt
package com.greentree.commons.math

import org.joml.Vector3f
import org.joml.Vector3fc
import org.joml.plus
import org.joml.unaryMinus
import java.util.*

class MathLine3D(
	point: Vector3fc,
	vector: Vector3fc,
) {

	private val point: Vector3fc = Vector3f(point)
	private var vector: Vector3fc = Vector3f(if(vector.x < 0) -vector else vector).normalize()

	fun p1(): Vector3fc {
		return vector
	}

	fun p2(): Vector3fc {
		return vector.plus(point)
	}

	override fun hashCode(): Int {
		return Objects.hash(point, vector)
	}

	override fun equals(obj: Any?): Boolean {
		if(this === obj) return true
		if(obj == null || javaClass != obj.javaClass) return false
		val other = obj as MathLine3D
		return point == other.point && vector == other.vector
	}

	override fun toString(): String {
		return "MathLine3D [vector=$vector, point=$point]"
	}
}
